pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -4,67 +4,243 @@ package querymiddleware
+		}
+	}
+
+	return r.next.RoundTrip(req)
+}
+
 // ParseRemoteReadRequestWithoutConsumingBody parses a remote read request
 // without consuming the body. It does not check the req.Body size, so it is
 // the caller's responsibility to ensure that the body is not too large.
 func ParseRemoteReadRequestWithoutConsumingBody(req *http.Request) (url.Values, error) {
-	params := make(url.Values)
+	remoteReadRequest, err := getRemoteReadRequestWithoutConsumingBody(req)
+	if err != nil {
+		return nil, err
+	}
+	return parseRemoteReadRequest(remoteReadRequest)
+}
 
+func getRemoteReadRequestWithoutConsumingBody(req *http.Request) (*prompb.ReadRequest, error) {
 	if req.Body == nil {
-		return params, nil
+		return nil, nil
 	}
 
 	bodyBytes, err := util.ReadRequestBodyWithoutConsuming(req)
 	if err != nil {
 		return nil, err
 	}
 
-	remoteReadRequest := prompb.ReadRequest{}
+	remoteReadRequest := &prompb.ReadRequest{}
 
-	_, err = util.ParseProtoReader(req.Context(), io.NopCloser(bytes.NewReader(bodyBytes)), int(req.ContentLength), querier.MaxRemoteReadQuerySize, nil, &remoteReadRequest, util.RawSnappy)
+	_, err = util.ParseProtoReader(req.Context(), io.NopCloser(bytes.NewReader(bodyBytes)), int(req.ContentLength), querier.MaxRemoteReadQuerySize, nil, remoteReadRequest, util.RawSnappy)
 	if err != nil {
 		return nil, err
 	}
 
+	return remoteReadRequest, nil
+}
+
+func parseRemoteReadRequest(remoteReadRequest *prompb.ReadRequest) (url.Values, error) {
+	if remoteReadRequest == nil {
+		return nil, nil
+	}
+
+	params := make(url.Values)
 	add := func(i int, name, value string) { params.Add(name+"_"+strconv.Itoa(i), value) }
 
 	queries := remoteReadRequest.GetQueries()
 
 	for i, query := range queries {
-		add(i, "start", fmt.Sprintf("%d", query.GetStartTimestampMs()))
-		add(i, "end", fmt.Sprintf("%d", query.GetEndTimestampMs()))
+		add(i, startLogKey, fmt.Sprintf("%d", query.GetStartTimestampMs()))
+		add(i, endLogKey, fmt.Sprintf("%d", query.GetEndTimestampMs()))
 
-		matchersStrings := make([]string, 0, len(query.Matchers))
-		matchers, err := remote.FromLabelMatchers(query.Matchers)
+		matcher, err := remoteReadMatchersToString(query)
 		if err != nil {
 			return nil, err
 		}
-		for _, m := range matchers {
-			matchersStrings = append(matchersStrings, m.String())
-		}
-		params.Add("matchers_"+strconv.Itoa(i), strings.Join(matchersStrings, ","))
+		add(i, matchersLogKey, matcher)
+
 		if query.Hints != nil {
 			if hints, err := json.Marshal(query.Hints); err == nil {
-				add(i, "hints", string(hints))
+				add(i, hintsLogKey, string(hints))
 			} else {
-				add(i, "hints", fmt.Sprintf("error marshalling hints: %v", err))
+				add(i, hintsLogKey, fmt.Sprintf("error marshalling hints: %v", err))
 			}
 		}
 	}
 
-	return params, err
+	return params, nil
+}
+
+func remoteReadMatchersToString(q *prompb.Query) (string, error) {
```

Comment:
[nit] Since they're the matchers of a single query:

```suggestion
func remoteReadQueryMatchersToString(q *prompb.Query) (string, error) {
```